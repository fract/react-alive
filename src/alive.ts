import React from 'react'
import { live, Fractal } from '@fract/core'

interface State {
    projection: null | JSX.Element
}

interface Props {
    target: Fractal<JSX.Element>
}

export class Alive extends React.Component<Props, State> {
    constructor(props: Props) {
        super(props)

        this.state = { projection: null }

        const update = (projection: JSX.Element) => this.setState({ projection })

        live(async function* () {
            while (true) yield update(yield* props.target)
        })
    }

    render() {
        return this.state.projection
    }
}
