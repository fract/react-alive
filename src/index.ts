import React from 'react'
import { Stream, whatsUp } from 'whatsup'

interface State {
    jsx: null | JSX.Element
}

interface Props {
    stream: Stream<JSX.Element>
}

export class WhatsUp extends React.Component<Props, State> {
    private dispose!: () => void

    constructor(props: Props) {
        super(props)

        this.state = { jsx: null }
    }

    componentDidMount() {
        const { stream } = this.props
        this.dispose = whatsUp(stream, (jsx) => this.setState({ jsx }))
    }

    componentWillUnmount() {
        this.dispose()
    }

    render() {
        return this.state.jsx
    }
}
