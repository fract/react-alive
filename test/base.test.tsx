import React from 'react'
import Renderer from 'react-test-renderer'
import { fractal, fraction } from '@fract/core'
import { Alive } from '../src/alive'

const delay = (time: number) => new Promise((r) => setTimeout(r, time))

const Name = fraction('John')

const User = fractal(async function* _User() {
    while (true) {
        yield <div>{yield* Name}</div>
    }
})

const renderer = Renderer.create(<Alive target={User} />)

it('initial render with null', async () => {
    const json = renderer.toJSON()
    expect(json).toMatchSnapshot()
})

it('render with <div>John</div>', async () => {
    await delay(300)
    const json = renderer.toJSON()
    expect(json).toMatchSnapshot()
})

it('render with <div>Barry</div>', async () => {
    Name.use('Barry')
    await delay(300)
    const json = renderer.toJSON()
    expect(json).toMatchSnapshot()
})
