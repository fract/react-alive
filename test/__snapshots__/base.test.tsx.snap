// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial render with null 1`] = `null`;

exports[`render with <div>Barry</div> 1`] = `
<div>
  Barry
</div>
`;

exports[`render with <div>John</div> 1`] = `
<div>
  John
</div>
`;
